napari/utils/_magicgui.py

```diff
@@ -16,21 +16,132 @@
 
 import weakref
 from functools import cache, partial
-from typing import TYPE_CHECKING, Any, get_args
+from typing import TYPE_CHECKING, Any, Optional, get_args
+
+import numpy as np
+from magicgui.widgets import ComboBox, FunctionGui
 
 from napari.utils._proxies import PublicOnlyProxy
 
 if TYPE_CHECKING:
     from concurrent.futures import Future
 
-    from magicgui.widgets import FunctionGui
-    from magicgui.widgets._bases import CategoricalWidget
+    from magicgui.widgets.bases import CategoricalWidget
 
     from napari._qt.qthreading import FunctionWorker
     from napari.layers import Layer
     from napari.viewer import Viewer
 
 
+def _get_layer_from_widget(gui: ComboBox, viewer: Viewer) -> Optional[Layer]:
+    """Retrieve layer used as input to function wrapped into magicgui .
+
+    Parameters
+    ----------
+    gui : magicgui.widgets.ComboBox
+        The instantiated ComboBox widget.
+
+    Returns
+    -------
+    Layer
+        layer passed as input to the function wrapped into magicgui.
+
+    """
+    import napari.types
+    from napari.layers import Layer
+
+    # this is workaround to work with python 3.9 as NewType do not contain
+    # information about mmodule
+    layer_types = {
```

Comment:
Since we've dropped Python 3.9, can we get rid of this block?

Reply:
yes